Use mutation creation time for hint expiry rather than hint submission time

Jollyplum opened the pull request:
patch by Matt Byrd; reviewed by Blake Eggleston for CASSANDRA-20014

https://issues.apache.org/jira/browse/CASSANDRA-20014

Commit:
Use mutation creation time for hint expiry rather than hint submission time

patch by Matt Byrd; reviewed by Chris Lohfink, Blake Eggleston for CASSANDRA-20014

clohfink commented on src/java/org/apache/cassandra/config/Config.java:
```diff
@@ -452,6 +452,7 @@ public static class SSTableConfig
     public ParameterizedClass hints_compression;
     public volatile boolean auto_hints_cleanup_enabled = false;
     public volatile boolean transfer_hints_on_decommission = true;
+    public volatile boolean hint_ttl_use_mutation_creation_time = true;
```
NP: maybe `use_creation_time_for_hint_ttl`

clohfink commented on test/distributed/org/apache/cassandra/distributed/test/HintDataReappearingTest.java:
```diff
@@ -0,0 +1,267 @@
+        doHintReappearData(false, true);
+    }
+
+
```
NP: remove blank lines

clohfink commented on test/distributed/org/apache/cassandra/distributed/test/HintDataReappearingTest.java:
```diff
@@ -0,0 +1,267 @@
+            transferHints(node1, node2hostId.call());
+
+            // Sleep a bit more to ensure hint is delivered after tombstone is GCed
+            Thread.sleep(20000);
```
20s is super long and gonna make this test a bit more of a pain, think we can slim this down some?

clohfink commented on src/java/org/apache/cassandra/service/StorageProxy.java:
```diff
@@ -2754,8 +2755,12 @@ public void runMayThrow()
-                HintsService.instance.write(hostIds, Hint.create(mutation, currentTimeMillis()));
+                long creationTime = currentTimeMillis();
+                if (DatabaseDescriptor.isHintTtlUseMutationCreationTime()) {
+                    creationTime -= TimeUnit.MILLISECONDS.convert(Math.abs(approxTime.now() - mutation.getApproxCreatedAtNanos()), NANOSECONDS);
```
would it make more sense to use the time from the PartitionUpdates DeletionTime inside the mutation instead of the time the mutation is created on the coordinator? I could see where that would be iffy since theres no guarantees client isn't using their own thing for writetimes.

clohfink reviewed (approved):
some minor questions and nitpicks but overall +1

Commit:
Address review comments change variable name, whitespace, capitlization and remove sleep in test

patch by Matt Byrd; reviewed by Chris Lohfink, Blake Eggleston for CASSANDRA-20014

bdeggleston commented on src/java/org/apache/cassandra/db/Mutation.java:
```diff
@@ -150,6 +150,11 @@ public ImmutableCollection<PartitionUpdate> getPartitionUpdates()
 
+    public long getApproxCreatedAtNanos()
+    {
+        return approxCreatedAtNanos;
```
`ModificationStatement#executeWithoutCondition` creates the write/delete timestamps before the mutations themselves are created. This means that hints could still be delivered late if enough time passes after creating the timestamps. 

This should err on the side of discarding hints early, so we should be recording a timestamp before actually creating mutation timestamps.

Jollyplum replied:
Yeah definitely agree that there is still this window(albeit likely much shorter) between mutation timestamp creation and the approximate time generated inside the mutation constructor.
At first glance it's quite a bit more fiddly to get some sort of timestamp pre mutation constructor timestamp and propagate it everywhere, but I can have a go at taking RequestTime.startedAtNanos or RequestTime.enqueuedAtNanos and plumbing it through to the mutation constructor everywhere. 

We'll also be adding a long in the Mutation object to solve this issue, so at fairly tiny but non-negligible cost.
Or I suppose slightly alternatively we could try and pull approxCreatedAtNanos out to just before the write/delete timestamps are generated, though this has some downstream effects due to it's other use, which need to be reasoned carefully about.

bdeggleston replied:
I probably wouldn't add it to Mutation. For `ModificationStatement#executeWithoutCondition` at least, you can grab nanoTime before the call to getMutations, then just pass it into the StorageProxy call as an argument. From there I think you need to plumb it to AbstractWriteResponseHandler and the call to `hintMutations`

bdeggleston reviewed (changes requested):
Could we also make maxHintTTL settable via JMX?

Jollyplum commented:
> Could we also make maxHintTTL settable via JMX?

Given it's a -D system property now, with a max int default, how did you want this to behave?
I see in some places we have -D that falls back on yaml, but they're only a couple of them and they look to be consumed just on startup and not settable via jmx e.g 
```
    public static String getAllocateTokensForKeyspace()
    {
        return ALLOCATE_TOKENS_FOR_KEYSPACE.getString(conf.allocate_tokens_for_keyspace);
```

I could make it perhaps a yaml property, which if the system property is set defaults to that? but it's breaking with quite a few conventions. Removing the -D and replacing with yaml only is probably also not feasible for something that's intended to be back-ported. 
Maybe allowing either to be set, but complaining if both are set on startup?

bdeggleston commented:
> Given it's a -D system property now, with a max int default, how did you want this to behave?

Initialization on startup wouldn't change, it would just become volatile and have some methods on StorageService or whatever to change it. I wouldn't add a yaml prop

Commit:
Make maxHintTTL be settable via jmx

patch by Matt Byrd; reviewed by Chris Lohfink, Blake Eggleston for CASSANDRA-20014

Commit:
Use request start time where available from write response handler, also
max the time offset with 0

patch by Matt Byrd; reviewed by Chris Lohfink, Blake Eggleston for CASSANDRA-20014

Commit:
add CHANGES.txt entry
patch by Matt Byrd; reviewed by Chris Lohfink, Blake Eggleston for CASSANDRA-20014

Jollyplum commented:
this was committed here:
https://github.com/apache/cassandra/commit/4f0ad22f122f8233aa99545bc234083690f53ea7
so closing the PR
